core/src/main/java/org/apache/iceberg/deletes/RoaringPositionBitmap.java

```diff
@@ -0,0 +1,317 @@
+ *
+ * <p>Positions must range from 0 (inclusive) to {@link #MAX_POSITION} (inclusive). The bitmap
+ * cannot handle positions exceeding {@link #MAX_POSITION} because the bitmap array size is a signed
+ * 32-bit integer, which must be non-negative. Allocating an array with size Integer.MAX_VALUE + 1
+ * would trigger an overflow and result in a negative length.
+ */
+class RoaringPositionBitmap {
+
+  static final long MAX_POSITION = toPosition(Integer.MAX_VALUE - 1, Integer.MIN_VALUE);
+  private static final RoaringBitmap[] EMPTY_BITMAP_ARRAY = new RoaringBitmap[0];
+  private static final long BITMAP_COUNT_SIZE_BYTES = 8L;
+  private static final long BITMAP_KEY_SIZE_BYTES = 4L;
+
+  private RoaringBitmap[] bitmaps;
+
+  RoaringPositionBitmap() {
+    this.bitmaps = EMPTY_BITMAP_ARRAY;
+  }
+
+  private RoaringPositionBitmap(RoaringBitmap[] bitmaps) {
+    this.bitmaps = bitmaps;
+  }
+
+  /**
+   * Adds a position to the bitmap.
+   *
+   * @param pos the row position
+   */
+  public void add(long pos) {
+    validatePosition(pos);
+    int key = key(pos);
+    int pos32Bits = pos32Bits(pos);
+    allocateBitmapsIfNeeded(key + 1 /* required number of bitmaps */);
+    bitmaps[key].add(pos32Bits);
+  }
+
+  /**
+   * Adds a range of positions to the bitmap.
+   *
+   * @param posStartInclusive the start position of the range (inclusive)
+   * @param posEndExclusive the end position of the range (exclusive)
+   */
+  public void addRange(long posStartInclusive, long posEndExclusive) {
+    for (long pos = posStartInclusive; pos < posEndExclusive; pos++) {
+      add(pos);
+    }
+  }
+
+  /**
+   * Adds all positions from the other bitmap to this bitmap, modifying this bitmap in place.
+   *
+   * @param that the other bitmap
+   */
+  public void addAll(RoaringPositionBitmap that) {
+    allocateBitmapsIfNeeded(that.bitmaps.length);
+    for (int key = 0; key < that.bitmaps.length; key++) {
+      bitmaps[key].or(that.bitmaps[key]);
+    }
+  }
+
+  /**
+   * Checks if the bitmap contains a position.
+   *
+   * @param pos the row position
+   * @return true if the position is in this bitmap, false otherwise
+   */
+  public boolean contains(long pos) {
+    validatePosition(pos);
+    int key = key(pos);
+    int pos32Bits = pos32Bits(pos);
+    return key < bitmaps.length && bitmaps[key].contains(pos32Bits);
+  }
+
+  /**
+   * Indicates whether the bitmap is empty.
+   *
+   * @return true if the bitmap is empty, false otherwise
+   */
+  public boolean isEmpty() {
+    return cardinality() == 0;
+  }
+
+  /**
+   * Returns the number of positions in the bitmap.
+   *
+   * @return the number of set positions
+   */
+  public long cardinality() {
+    long cardinality = 0L;
+    for (RoaringBitmap bitmap : bitmaps) {
+      cardinality += bitmap.getLongCardinality();
+    }
+    return cardinality;
+  }
+
+  /**
+   * Compresses the bitmap wherever it is more space efficient.
+   *
+   * @return whether the bitmap was compressed
```

Comment:
Nit:
I think i have a grammar issue with this in my head. Since the bitmap is always compressed when it comes out of this method. If it's false then the bitmap "was not compressed" but it "is compressed". 

I can't think of anything better though ... 
@return false if this was a no-op, true if it was changed
That's also not good....

Reply:
I renamed this part a bit.